packages/vite/src/module-runner/runner.ts

```diff
@@ -64,17 +68,24 @@ export class ModuleRunner {
     const root = this.options.root
     this.root = root[root.length - 1] === '/' ? root : `${root}/`
     this.evaluatedModules = options.evaluatedModules ?? new EvaluatedModules()
-    this.transport = options.transport
+    this.transport = normalizeModuleRunnerTransport(options.transport)
+    let hmrHandlerForTransport: ((payload: HotPayload) => void) | undefined
     if (typeof options.hmr === 'object') {
-      this.hmrClient = new HMRClient(
+      const resolvedHmrLogger =
         options.hmr.logger === false
           ? silentConsole
-          : options.hmr.logger || hmrLogger,
-        options.hmr.connection,
+          : options.hmr.logger || hmrLogger
+      this.hmrClient = new HMRClient(
+        resolvedHmrLogger,
+        this.transport,
         ({ acceptedPath }) => this.import(acceptedPath),
       )
-      options.hmr.connection.onUpdate(createHMRHandler(this))
+      hmrHandlerForTransport = createHMRHandler(this)
+      if (!this.transport.connect) {
+        resolvedHmrLogger.error('HMR is not supported by this runner transport')
```

Comment:
Should we throw an error instead? 🤔 If `hmr` is provided, then I would assume the transport supports it

Reply:
I changed the way in f08a2dfddba3415c92074d904a9de5db0216d7f5 and fddfd6494821fdd4e24d2c814cdcf5a7aaf06062, along with the following comments.